### DISCUSSION TITLE
Query all selectors on one page

### DISCUSSION THREAD

**Discussion by ghost** [typst/typst#4449] [Q&A] [answered]
I am typesetting a dictionary. The header should contain the first and last headwords of each page.
Headwords are level 3 headings in my scheme. The following works for displaying the first headword:
```typst
#set page(
  header: context {
    let headwords = query(
      selector(heading.where(level: 3)).after(here()),
    )
    if headwords.len() == 0 {
    } else  {
      headwords.first().body
    }
  }
)
``` 
But that works by querying an array of all headings after the current location on the page, till the end of the document. A tad overkill, and it doesn't help getting the last heading on that one page.

**Comment by l0wx** [comment, marked as answer]
`….before(selector(<used-in-footer>).after(here()))` or just `query(…).filter(h3=> here().page() == h3.location().page())`

**Reply by ghost** [reply]
Thank you very much. I wasn't sure how to use `location()`. Here is is the code, with a provision for the case when there is only one headword on the page:
```typst
#set page(
   header: context {
     let headwords = query(
       selector(heading.where(level: 3)))
       .filter(h3 => here().page() == h3.location().page())
     if headwords.len() == 0 {
     } else if headwords.last() == headwords.first() {
       headwords.first().body
     } else {
       headwords.first().body + " " + headwords.last().body
    } 
   }
 )
```
